onnxruntime/core/session/inference_session.cc

```diff
@@ -415,96 +415,72 @@ class InferenceSession::Impl {
                   "Unexpected input data type. Actual: (" + actual_name + ") , expected: (" + expected_name + ")");
   }
 
-  common::Status ValidateInputTypes(const std::vector<std::string>& feed_names,
-                                    const std::vector<MLValue>& feeds) {
+  common::Status ValidateInputs(const std::vector<std::string>& feed_names,
+                                const std::vector<MLValue>& feeds) {
     const auto begin_names = feed_names.cbegin();
     const auto end_names = feed_names.cend();
-    for (auto& arg : input_def_list_) {
+    std::unordered_set<ptrdiff_t> required_feed_ids;
+    for (auto& arg : required_input_def_list_) {
       auto& arg_name = arg->Name();
       if (arg_name.empty()) {
         continue;
       }
 
       auto feed_names_entry = std::find(begin_names, end_names, arg_name);
       if (feed_names_entry == end_names) {
-        continue;
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                               "Missing required input: ", arg_name);
       }
 
       auto idx = feed_names_entry - begin_names;
+      required_feed_ids.insert(idx);
       auto& input_ml_value = feeds.at(idx);
-      auto input_type = input_ml_value.Type();
       auto expected_type = utils::GetMLDataType(*arg);
 
-      if (!input_ml_value.IsTensor()) {
-        auto retval = CheckTypes(input_type, expected_type);
-        if (!retval.IsOK()) {
-          return retval;
-        }
-        continue;
-      }
-
-      auto expected_element_type = expected_type->AsTensorType()->GetElementType();
-      auto input_element_type = input_ml_value.Get<Tensor>().DataType();
-      auto retval = CheckTypes(input_element_type, expected_element_type);
-      if (!retval.IsOK()) {
-        return retval;
+      if (input_ml_value.IsTensor()) {
+        auto expected_element_type = expected_type->AsTensorType()->GetElementType();
+        auto input_element_type = input_ml_value.Get<Tensor>().DataType();
+        ORT_RETURN_IF_ERROR(CheckTypes(input_element_type, expected_element_type));
+      } else {
+        auto input_type = input_ml_value.Type();
+        ORT_RETURN_IF_ERROR(CheckTypes(input_type, expected_type));
       }
     }
-    return Status::OK();
-  }
 
-  common::Status ValidateInputNames(const std::vector<std::string>& feed_names) {
-    std::string missing_required_inputs;
+    if (feeds.size() > required_feed_ids.size()) {
+      // More feeds are offered.
+      // In the case of overriding some initializers (which are also taken as graph inputs).
+      for (size_t i = 0; i < feeds.size(); ++i) {
```

Comment:
Just for clarity given this contradicts the original "we don't allow overwriting weights" statement, is this a short term thing to enable the truncated sequence use cases (and they will update the model to remove the initializers with defaults for the previous iteration values and pass a full set of inputs on every call), or permanent?

Reply:
it's not a short term thing, and we'll allow this overwriting with default value for graph inputs.